docs/proposals/api-group.md

```diff
@@ -0,0 +1,135 @@
+
+  For backward compatibility, v1 objects belong to the group with an empty name, so existing v1 config files will remain valid.
+
+2. conversion.Scheme:
+
+  The key of conversion.Scheme.versionMap will be "group/version". The internal version of objects of all groups will be registered to versionMap[""].
+
+3. meta.RESTMapper:
+
+  Each group will have its own RESTMapper, and these mappers will be registered to api.RESTMapper.
+
+  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified and there is ambiguity (i.e., the resource exists in multiple groups), an error should be returned to force the user to specify the group.
+
+## Server-side implementation
+
+1. resource handlers' URL:
+
+  We will force the URL to be in the form of prefix/group/version/...
+
+    * for api/v1, we set the prefix="api" (which is populated from cmd/kube-apiserver/app#APIServer.APIPrefix), group="", version="v1", so the URL remains to be /api/v1
+
+    * for new kube API groups, we will set the prefix="kube-api" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /kube-api/experimental/v1alpha1
+
+    * for OpenShift and other third party API, they should be able to name their own prefix and group. For example, OpenShift can either choose prefix="osapi", group="", or prefix="", group="osapi", so the URL remains the same as it is now.
+
+  Prefix is used to differentiate API paths from other paths like \healthz.
+
+2. exposing supported API groups:
+
+  API Server will expose the supported groups and versions in the form of "group/version" at prefix/ (e.g., kube-api/). The Swagger spec should be exposed at /swaggerspec/prefix/group/version/ (e.g., /swaggerspec/kube-api/experimental/v1alpha1).
+
+
+3. handling common API objects:
+
+  * top-level common API objects like api.Status:
+
+    To handle the top-level API objects that are used by all groups, we either have to register them to all schemes, or we can choose not to encode them to a version.
+
+    Currently, for api.Status, we choose the latter approach: we will move it from pkg/api/types.go to pkg/api/unversioned.go; for backward compatibility, the server will encode it if the client expected the Status to be encoded in v1, otherwise the server will send the internal version, i.e., api.Status. If an error occurs before the version can be determined, the server will send the internal version.
+
+  * non-top-level common API objects:
+
+    Assuming object x belongs to group X is used as a field in an object belongs to group Y, currently genconversion will generate the conversion functions for x in package Y. Hence, we don't need any special treatment for non-top-level common API objects.
```

Comment:
This sentence reads a little awkwardly to me.  Do you mean:

`Assuming object o belonging to group X is used as a field in an object belonging to group Y`